OpenRA.Game/Network/Connection.cs

```diff
@@ -153,27 +154,27 @@ protected virtual void Send(byte[] packet)
 		{
 			if (packet.Length == 0)
 				throw new NotImplementedException();
+
 			AddPacket(new ReceivedPacket { FromClient = LocalClientId, Data = packet });
 		}
 
 		protected void AddPacket(ReceivedPacket packet)
 		{
-			lock (receivedPackets)
-				receivedPackets.Add(packet);
+			receivedPackets.Enqueue(packet);
 		}
 
-		public virtual void Receive(Action<int, byte[]> packetFn)
+		public virtual void Receive(Action<int, byte[], int> packetFn)
 		{
-			ReceivedPacket[] packets;
-			lock (receivedPackets)
+			var packets = new List<ReceivedPacket>(receivedPackets.Count);
+
+			while (receivedPackets.TryDequeue(out var received))
 			{
-				packets = receivedPackets.ToArray();
-				receivedPackets.Clear();
+				packets.Add(received);
 			}
 
 			foreach (var p in packets)
 			{
-				packetFn(p.FromClient, p.Data);
+				packetFn(p.FromClient, p.Data, p.Timestep);
```

Comment:
I really don't understand what is going on with passing this timestamp through all the order plumbing. Why isn't this encoded inside  `p.Data` and dealt with at the higher level parsing?